File: clients/py/src/seismic_web3/transaction/metadata.py
"""Seismic transaction metadata builder.

Builds ``TxSeismicMetadata`` by fetching chain state (nonce, latest
block) as needed, with both sync and async variants.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from seismic_web3.crypto.nonce import random_encryption_nonce

if TYPE_CHECKING:
    from eth_typing import ChecksumAddress
    from web3 import AsyncWeb3, Web3

    from seismic_web3._types import Bytes32, CompressedPublicKey, EncryptionNonce
    from seismic_web3.transaction_types import TxSeismicMetadata

#: Default number of blocks before a transaction expires.
DEFAULT_BLOCKS_WINDOW = 100


@dataclass
class MetadataParams:
    """Parameters for building ``TxSeismicMetadata``.

    Optional fields (``None``) are automatically fetched from the chain
    or generated.

    Attributes:
        sender: Checksummed sender address.
        to: Recipient address (``None`` for contract creation).
        encryption_pubkey: Compressed public key for ECDH.
        value: Wei to transfer (default ``0``).
        nonce: Sender's tx count (fetched if ``None``).
        blocks_window: Blocks until expiry (default ``100``).
        encryption_nonce: AES-GCM nonce (random if ``None``).
        recent_block_hash: Recent block hash (fetched if ``None``).
        expires_at_block: Explicit expiry block (computed if ``None``).
        message_version: ``0`` for raw, ``2`` for EIP-712.
        signed_read: ``True`` for signed ``eth_call`` reads.
    """

    sender: ChecksumAddress
    to: ChecksumAddress | None
    encryption_pubkey: CompressedPublicKey
    value: int = 0
    nonce: int | None = None
    blocks_window: int = DEFAULT_BLOCKS_WINDOW
    encryption_nonce: EncryptionNonce | None = None
    recent_block_hash: Bytes32 | None = None
    expires_at_block: int | None = None
    message_version: int = 0
    signed_read: bool = False


def _assemble_metadata(
    params: MetadataParams,
    chain_id: int,
    nonce: int,
    recent_block_hash: Bytes32,
    expires_at_block: int,
    encryption_nonce: EncryptionNonce,
) -> TxSeismicMetadata:
    """Pure helper: assemble metadata from resolved values."""
    from seismic_web3.transaction_types import (
        LegacyFields,
        SeismicElements,
        TxSeismicMetadata,
    )

    return TxSeismicMetadata(
        sender=params.sender,
        legacy_fields=LegacyFields(
            chain_id=chain_id,
            nonce=nonce,
            to=params.to,
            value=params.value,
        ),
        seismic_elements=SeismicElements(
            encryption_pubkey=params.encryption_pubkey,
            encryption_nonce=encryption_nonce,
            message_version=params.message_version,
            recent_block_hash=recent_block_hash,
            expires_at_block=expires_at_block,
            signed_read=params.signed_read,
        ),
    )


def build_metadata(w3: Web3, params: MetadataParams) -> TxSeismicMetadata:
    """Build ``TxSeismicMetadata``, fetching chain state as needed (sync).

    Resolves ``nonce``, ``recent_block_hash``, and ``expires_at_block``
    from the connected node if not explicitly provided in ``params``.

    Args:
        w3: Sync ``Web3`` instance.
        params: Metadata parameters (some may be ``None``).

    Returns:
        Fully populated ``TxSeismicMetadata``.
    """
    from seismic_web3._types import Bytes32

    chain_id = w3.eth.chain_id
    nonce = (
        params.nonce
        if params.nonce is not None
        else w3.eth.get_transaction_count(params.sender)
    )
    enc_nonce = params.encryption_nonce or random_encryption_nonce()

    if params.recent_block_hash is not None and params.expires_at_block is not None:
        block_hash = params.recent_block_hash
        expires = params.expires_at_block
    else:
        block = w3.eth.get_block("latest")
        block_hash = params.recent_block_hash or Bytes32(block["hash"])
        if params.expires_at_block is not None:
            expires = params.expires_at_block
        else:
            expires = block["number"] + params.blocks_window

    return _assemble_metadata(params, chain_id, nonce, block_hash, expires, enc_nonce)


async def async_build_metadata(
    w3: AsyncWeb3, params: MetadataParams
) -> TxSeismicMetadata:
    """Build ``TxSeismicMetadata``, fetching chain state as needed (async).

    Async variant of :func:`build_metadata`.

    Args:
        w3: Async ``AsyncWeb3`` instance.
        params: Metadata parameters (some may be ``None``).

    Returns:
        Fully populated ``TxSeismicMetadata``.
    """
    from seismic_web3._types import Bytes32

    chain_id = await w3.eth.chain_id
    nonce = (
        params.nonce
        if params.nonce is not None
        else await w3.eth.get_transaction_count(params.sender)
    )
    enc_nonce = params.encryption_nonce or random_encryption_nonce()

    if params.recent_block_hash is not None and params.expires_at_block is not None:
        block_hash = params.recent_block_hash
        expires = params.expires_at_block
    else:
        block = await w3.eth.get_block("latest")
        block_hash = params.recent_block_hash or Bytes32(block["hash"])
        if params.expires_at_block is not None:
            expires = params.expires_at_block
        else:
            expires = block["number"] + params.blocks_window

    return _assemble_metadata(params, chain_id, nonce, block_hash, expires, enc_nonce)

File: clients/ts/packages/seismic-viem/src/tx/metadata.ts
/**
 * Seismic tx metadata: the inputs needed to construct, encrypt, and sign a
 * Seismic transaction beyond what a caller normally provides.
 *
 * A plain viem `sendTransaction` caller supplies `to`, `value`, `data`, and
 * optionally `nonce` / gas fields. A Seismic tx additionally requires:
 *
 *  - the client's encryption pubkey + a fresh encryption nonce (for AEAD)
 *  - a `recentBlockHash` + `expiresAtBlock` window (to bound replay)
 *  - a `messageVersion` selecting the signing path
 *  - a `signedRead` flag (true for signed eth_call, false for txs)
 *  - the resolved tx nonce and chainId (to bind the AAD)
 *
 * `TxSeismicMetadata` bundles all of the above so that tx-build paths
 * (`sendShielded`, `signedCall`, smart contract writes) can share one
 * resolution routine and one AAD encoding (see `crypto/aead.ts`).
 */
import {
  Account,
  Address,
  Chain,
  GetBlockParameters,
  GetTransactionCountParameters,
  Transport,
} from 'viem'
import { parseAccount } from 'viem/accounts'

import { ShieldedWalletClient } from '@sviem/client.ts'
import { randomEncryptionNonce } from '@sviem/crypto/nonce.ts'
import type {
  SeismicBlockParams,
  SeismicElements,
  SeismicSecurityParams,
} from '@sviem/tx/seismicTx.ts'
import { TYPED_DATA_MESSAGE_VERSION } from '@sviem/tx/signSeismicTypedData.ts'

const DEFAULT_SEISMIC_BLOCKS_WINDOW = 100n

/**
 * Ethereum-compatible fields pulled out of the metadata so they can be bound
 * into the AEAD Additional Authenticated Data. Keeping them in their own
 * sub-object makes the AAD encoding explicit and symmetric with the Rust
 * reference implementation in the node.
 */
export type LegacyFields = {
  chainId: number
  nonce: number
  to: Address | null | undefined
  value: bigint
}

/**
 * Resolved metadata for a single Seismic transaction or signed read.
 *
 * Consumers:
 *  - `sendShielded` / `signedCall` — pass this straight into the encryption
 *    step and the tx builder
 *  - `encryptionActions.encrypt/decrypt` — uses the metadata to derive the AAD
 *    bound to the ciphertext
 *  - `signSeismicTxTypedData` — reads `seismicElements` to populate the
 *    EIP-712 message
 *
 * Build instances with `buildTxSeismicMetadata(...)`.
 */
export type TxSeismicMetadata = {
  sender: Address
  legacyFields: LegacyFields
  seismicElements: SeismicElements
}

const fillNonce = async <
  TChain extends Chain | undefined,
  TAccount extends Account,
>(
  client: ShieldedWalletClient<Transport, TChain, TAccount>,
  parameters: {
    account: Account | null
    nonce: number | undefined
  } & GetBlockParameters
) => {
  let account = parseAccount(parameters.account || client.account)
  const { nonce: nonce_ } = parameters
  if (nonce_ !== undefined) {
    return nonce_
  }

  const { blockNumber, blockTag = 'pending' } = parameters
  let args: GetTransactionCountParameters = {
    address: account.address,
    blockTag,
  }
  if (blockNumber) {
    args = { address: account.address, blockNumber }
  }
  return client.getTransactionCount(args)
}

type BuildTxSeismicMetadataParams = {
  account: Address | Account | null | undefined
  nonce?: number
  to: Address
  value?: bigint
  typedDataTx?: boolean
  signedRead?: boolean
} & SeismicSecurityParams

const inferTypedDataTx = (
  typedDataTx: boolean | undefined,
  account: Account
): boolean => {
  if (typedDataTx !== undefined) {
    return typedDataTx
  }
  return account.type === 'json-rpc' || account.type === 'local'
}

/**
 * Resolves a `TxSeismicMetadata` for a tx or signed read.
 *
 * Fills the Seismic-specific fields that a caller rarely wants to compute by
 * hand:
 *
 *  - `nonce` — fetched via `getTransactionCount({ blockTag: 'pending' })` if
 *    not supplied
 *  - `encryptionNonce` — randomly generated if not supplied
 *  - `recentBlockHash` / `expiresAtBlock` — defaults to the latest block and
 *    `latest.number + blocksWindow` (default window: 100 blocks)
 *  - `messageVersion` — `TYPED_DATA_MESSAGE_VERSION` for json-rpc/local
 *    accounts (so wallets like MetaMask can sign via EIP-712), `0` otherwise
 *
 * Any of the above can be forced via `SeismicSecurityParams` — mainly used
 * by tests that need deterministic metadata or callers that want explicit
 * expiry control.
 *
 * Throws if `expiresAtBlock` is in the past or if `blocksWindow <= 0`.
 */
export const buildTxSeismicMetadata = async <
  TChain extends Chain | undefined,
  TAccount extends Account,
>(
  client: ShieldedWalletClient<Transport, TChain, TAccount>,
  {
    account: paramsAcct,
    nonce,
    to,
    value = 0n,
    encryptionNonce,
    blocksWindow = DEFAULT_SEISMIC_BLOCKS_WINDOW,
    recentBlockHash,
    expiresAtBlock,
    typedDataTx,
    signedRead = false,
  }: BuildTxSeismicMetadataParams
): Promise<TxSeismicMetadata> => {
  const account = parseAccount(paramsAcct || client.account)
  if (!account) {
    throw new Error(`Signed reads must have an account`)
  }
  if (blocksWindow <= 0n) {
    throw new Error(`blocksWindow param must be > 0`)
  }

  const resolveBlockParams = async (): Promise<SeismicBlockParams> => {
    if (recentBlockHash && expiresAtBlock) {
      return { recentBlockHash, expiresAtBlock }
    }
    if (recentBlockHash) {
      const recentBlock = await client.getBlock({ blockHash: recentBlockHash })
      return {
        recentBlockHash: recentBlock.hash,
        expiresAtBlock: recentBlock.number + blocksWindow,
      }
    }
    const latestBlock = await client.getBlock({ blockTag: 'latest' })
    if (expiresAtBlock) {
      if (expiresAtBlock <= latestBlock.number) {
        throw new Error(
          `expiresAtBlock param ${expiresAtBlock} is in the past (latest block is #${latestBlock.number})`
        )
      }
      return { recentBlockHash: latestBlock.hash, expiresAtBlock }
    }
    return {
      recentBlockHash: latestBlock.hash,
      expiresAtBlock: latestBlock.number + blocksWindow,
    }
  }

  const useTypedDataTx = inferTypedDataTx(typedDataTx, account)
  const [nonce_, chainId, blockParams] = await Promise.all([
    fillNonce(client, { account, nonce }),
    client.getChainId(),
    resolveBlockParams(),
  ])

  if (client.chain) {
    if (client.chain.id !== chainId) {
      throw new Error(`Client chain's id does not match eth_chainId response`)
    }
  }

  return {
    sender: account.address,
    legacyFields: {
      chainId,
      nonce: nonce_,
      to,
      value,
    },
    seismicElements: {
      encryptionPubkey: client.getEncryptionPublicKey(),
      encryptionNonce: encryptionNonce ?? randomEncryptionNonce(),
      messageVersion: useTypedDataTx ? TYPED_DATA_MESSAGE_VERSION : 0,
      recentBlockHash: blockParams.recentBlockHash,
      expiresAtBlock: blockParams.expiresAtBlock,
      signedRead,
    },
  }
}
